parity/db/rocksdb/migration_secretstore.rs

```diff
@@ -0,0 +1,160 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Secret Store DB migration module.
+
+
+use std::fmt::{Display, Error as FmtError, Formatter};
+use std::fs;
+use std::io::{Error as IoError, ErrorKind as IoErrorKind, Read as _, Write as _};
+use std::path::PathBuf;
+
+use kvdb::DBTransaction;
+use super::kvdb_rocksdb::{Database, DatabaseConfig};
+
+/// We used to store the version in the database (until version 4).
+const LEGACY_DB_META_KEY_VERSION: &'static [u8; 7] = b"version";
+/// Current db version.
+const CURRENT_VERSION: u8 = 4;
+/// Database is assumed to be at the default version, when no version file is found.
+const DEFAULT_VERSION: u8 = 3;
+/// Version file name.
+const VERSION_FILE_NAME: &'static str = "db_version";
+
+/// Migration related erorrs.
+#[derive(Debug)]
+pub enum Error {
+	/// Returned when current version cannot be read or guessed.
+	UnknownDatabaseVersion,
+	/// Existing DB is newer than the known one.
+	FutureDBVersion,
+	/// Migration was completed succesfully,
+	/// but there was a problem with io.
+	Io(IoError),
+}
+
+impl Display for Error {
+	fn fmt(&self, f: &mut Formatter) -> Result<(), FmtError> {
+		let out = match *self {
+			Error::UnknownDatabaseVersion =>
+				"Current Secret Store database version cannot be read".into(),
+			Error::FutureDBVersion =>
+				"Secret Store database was created with newer client version.\
+				Upgrade your client or delete DB and resync.".into(),
+			Error::Io(ref err) =>
+				format!("Unexpected io error on Secret Store database migration: {}.", err),
+		};
+		write!(f, "{}", out)
+	}
+}
+
+impl From<IoError> for Error {
+	fn from(err: IoError) -> Self {
+		Error::Io(err)
+	}
+}
+
+// Moved "default" column to column 0 in preparation for a kvdb-rocksdb 0.3 migration.
+fn migrate_to_v4(parent_dir: &str) -> Result<(), Error> {
+	// Naïve implementation until
+	// https://github.com/facebook/rocksdb/issues/6130 is resolved
+	let old_db_config = DatabaseConfig {
+		columns: None,
+		..Default::default()
+	};
+	let new_db_config = DatabaseConfig {
+		columns: Some(1),
+		..Default::default()
+	};
+	const BATCH_SIZE: usize = 1024;
+	let new_dir = migration_dir(parent_dir);
+	let old_db = Database::open(&old_db_config, &db_dir(parent_dir))?;
+	let new_db = Database::open(&new_db_config, &new_dir)?;
+
+	// remove legacy version key
+	{
+		let mut batch = DBTransaction::with_capacity(1);
+		batch.delete(None, LEGACY_DB_META_KEY_VERSION);
+		old_db.write(batch)?;
```

Comment:
Worth logging here imo (I don't think the error from `write` is very specific): `error!("db version deletion failed: {}")` or something?